File: src/toolchain/host/platforms/ios/Sources/Z42VM/ZpkgResolver.swift
import Foundation

/// Resolve a namespace (e.g. `"z42.core"`, `"Std.IO"`) to its zpkg
/// bytes. Returning `nil` signals "this resolver doesn't know about
/// that namespace"; the runtime then falls back to the next configured
/// resolver (typically a search-path scan).
///
/// Spec: docs/design/runtime/embedding.md §11.
public protocol ZpkgResolver {
    func resolve(namespace: String) -> Data?
}

/// Default iOS resolver. Resolution order:
///   1. **Namespace index** (`<subdirectory>/index.json`) — map
///      `"Std.IO" → "z42.io.zpkg"` etc. Index is produced by host
///      `scripts/build-stdlib.sh` and shipped via `build.sh`.
///   2. **Filename fallback** — namespace-as-filename
///      (`<subdirectory>/<namespace>.zpkg`); useful for custom resolvers
///      that publish single-namespace-per-file packages without an index.
///
/// Index makes the resolver correct for stdlib where one zpkg provides
/// multiple namespaces (e.g. `z42.core.zpkg` ships `Std` + `Std.Exceptions`).
/// Fallback preserves the simple "one namespace per file" convention for
/// hosts that don't ship an index.
///
/// Spec: docs/spec/archive/2026-05-12-fix-bundle-resolver-namespace-index/
public struct BundleZpkgResolver: ZpkgResolver {
    public let bundle: Bundle
    public let subdirectory: String?
    private let index: [String: String]

    public init(bundle: Bundle = .main, subdirectory: String? = "stdlib") {
        self.bundle = bundle
        self.subdirectory = subdirectory
        self.index = Self.loadIndex(bundle: bundle, subdirectory: subdirectory)
    }

    public func resolve(namespace: String) -> Data? {
        // 1. Index-driven lookup.
        if let filename = index[namespace],
           let data = readResource(filename: filename) {
            return data
        }
        // 2. Filename fallback (namespace-as-basename).
        return readResource(filename: "\(namespace).zpkg")
    }

    /// Load `<subdirectory>/<filename>` (e.g. `stdlib/z42.io.zpkg`) and
    /// return its bytes; nil on any error (file missing / unreadable).
    private func readResource(filename: String) -> Data? {
        let nsFilename = filename as NSString
        let basename = nsFilename.deletingPathExtension
        let ext = nsFilename.pathExtension
        guard !basename.isEmpty, !ext.isEmpty,
              let url = bundle.url(
                  forResource: basename,
                  withExtension: ext,
                  subdirectory: subdirectory
              )
        else { return nil }
        return try? Data(contentsOf: url)
    }

    private static func loadIndex(bundle: Bundle, subdirectory: String?) -> [String: String] {
        guard let url = bundle.url(
            forResource: "index",
            withExtension: "json",
            subdirectory: subdirectory
        ),
        let data = try? Data(contentsOf: url),
        let parsed = try? JSONSerialization.jsonObject(with: data, options: []),
        let map = parsed as? [String: String]
        else { return [:] }
        return map
    }
}

/// HashMap-backed resolver useful for tests and for hosts that build
/// the zpkg dictionary at startup (e.g. fetched over network).
public final class MapZpkgResolver: ZpkgResolver {
    private var map: [String: Data]

    public init(_ initial: [String: Data] = [:]) {
        self.map = initial
    }

    public func set(_ namespace: String, _ bytes: Data) {
        map[namespace] = bytes
    }

    public func resolve(namespace: String) -> Data? {
        map[namespace]
    }
}

File: src/toolchain/host/platforms/android/z42vm/src/main/java/io/z42/vm/ZpkgResolver.kt
package io.z42.vm

import android.content.res.AssetManager
import java.io.IOException

/**
 * Resolve a namespace (e.g. `"z42.core"`, `"Std.IO"`) to its zpkg
 * bytes. Returning `null` signals "this resolver doesn't know about
 * that namespace"; the runtime then falls back to other resolvers
 * (typically a search-path scan, which Android doesn't use).
 *
 * Spec: docs/design/runtime/embedding.md §11.
 */
interface ZpkgResolver {
    fun resolve(namespace: String): ByteArray?
}

/**
 * Default Android resolver. Resolution order:
 *   1. **Namespace index** (`<subdir>/index.json`) — map
 *      `"Std.IO" → "z42.io.zpkg"` etc. Index is produced by host
 *      `scripts/build-stdlib.sh` and shipped via `build.sh`.
 *   2. **Filename fallback** — namespace-as-filename
 *      (`<subdir>/<namespace>.zpkg`); preserves the simple
 *      "one namespace per file" convention for hosts that don't ship
 *      an index.
 *
 * Index makes the resolver correct for stdlib where one zpkg provides
 * multiple namespaces (e.g. `z42.core.zpkg` ships `Std` + `Std.Exceptions`).
 *
 * Spec: docs/spec/archive/2026-05-12-fix-bundle-resolver-namespace-index/
 */
class AssetZpkgResolver(
    private val assets: AssetManager,
    private val subdir: String = "stdlib",
) : ZpkgResolver {
    private val index: Map<String, String> = loadIndex()

    override fun resolve(namespace: String): ByteArray? {
        // 1. Index-driven lookup.
        index[namespace]?.let { filename ->
            try {
                return assets.open("$subdir/$filename").use { it.readBytes() }
            } catch (_: IOException) {
                /* fall through to filename fallback */
            }
        }
        // 2. Filename fallback (namespace-as-basename).
        return try {
            assets.open("$subdir/$namespace.zpkg").use { it.readBytes() }
        } catch (_: IOException) {
            null
        }
    }

    private fun loadIndex(): Map<String, String> = try {
        val text = assets.open("$subdir/index.json").use {
            it.readBytes().toString(Charsets.UTF_8)
        }
        val obj = org.json.JSONObject(text)
        val result = HashMap<String, String>(obj.length())
        val keys = obj.keys()
        while (keys.hasNext()) {
            val k = keys.next()
            result[k] = obj.getString(k)
        }
        result
    } catch (_: Exception) {
        emptyMap()
    }
}

/**
 * HashMap-backed resolver useful for tests and for hosts that build
 * the zpkg dictionary at startup (e.g. fetched over network).
 */
class MapZpkgResolver(
    initial: Map<String, ByteArray> = emptyMap(),
) : ZpkgResolver {
    private val map = HashMap(initial)

    fun set(namespace: String, bytes: ByteArray) {
        map[namespace] = bytes
    }

    override fun resolve(namespace: String): ByteArray? = map[namespace]
}
